### PR TITLE
Add AppSync API

### PR DESCRIPTION
**Pull request by bboure**
Guidelines:

# resolver definition:

In the `graphql.yml` file, under the `mappingTemplates`, is where you can define resolvers:

- `type` is the parent graphql type (where the field is located). eg: `Query` or `Organization`
- `field` is the field for this resolver
- `dataSource` is source where the data comes from (more on that later)
- `request` and `response` are the VTL templates running before/after the request. 
For Lambda, I recommend not using them at all (use [direct lambda](https://aws.amazon.com/blogs/mobile/appsync-direct-lambda/)). For that, set `false` as request/response
For the rest of data sources (like DynamoDb), don't specify anything. By default, you can write the VTL at the following location: `Type.field.request.vtl` and `Type.field.response.vtl` (see [example](https://github.com/RareCamp/RareCamp/pull/15/files#diff-30ad8db641ed38e640bbd3fe883c0793e2ef13983a04ea79dd184826077d2701))

# Data sources

Each data source must be declared under the `dataSources` property (graphql.yml file).
Sources can be `AMAZON_DYNAMODB`, `AWS_LAMBDA`, `HTTP`, `ELASTICSEARCH` or `AWS_RDS`. You should be good with only DynamoDb and Lambda.
Check thee examples on how to configure them.
Lambda functions must be also created in the `fucntions.yml` file
DynamoDb tables must exist or be created in the `dynamodb.yml` files

# testing
Check [this repo](https://github.com/theburningmonk/appsyncmasterclass-backend/tree/main/__tests__/test_cases/unit) for good testing use cases

# Offline support
When working locally, or for testing locally, without deploying, I pre-installed [serverless-appsync-simulator](https://github.com/bboure/serverless-appsync-simulator)

Usage:

```bash
sls offline start
```

You will need to install DynamoDb and run migrations first (only once)

```bash
sls dynamodb install
sls dynamodb migrate
```

* Run `sls dynamodb migrate` again if you add new talbes.

-------------------------

Any other question, please ping me

### CODE REVIEW

File: serverless.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ provider:
   name: aws
   stackName: ${{self:service}}-${{self:provider.stage}}
   runtime: nodejs12.x
-  memorySize: 1024
+  memorySize: 512
```
Comment (sanathkr) [inline serverless.yaml:6]: Great call out! Will save us some good money!

File: templates/dynamodb.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+Resources:
+  UserTable:
```
Comment (sanathkr) [inline templates/dynamodb.yml:2]: This is pretty cool. I didn't know we could break them out like this into multiple files!

**Comment by pipesorter** [comment]
@bboure 👍  I have a question, do you suggest using  the new `Datastore` https://docs.amplify.aws/lib/datastore/getting-started/q/platform/js#sample-schema in our react app or do we have to use make graphql queries in combination of something like `react-query` ?

**Comment by bboure** [comment]
@pipesorter I am not using Amplify, so I can't really tell. (I personally use the [Apollo Client](https://www.apollographql.com/docs/react/))

The good thing about Amplify, however, is that it handles offline support, Cognito, etc., out-of-the-box.
I would definitely consider it for this project.

File: serverless.yaml (hunk 1 above)
Comment (bboure) [reply serverless.yaml:6]: In my experience, 512 is usually a good compromise between cost and performance. 
For the future, there is [a tool](https://www.npmjs.com/package/lumigo-cli#lumigo-cli-analyze-lambda-cost) that analyses cost and tells you the best match for each individual function

File: mapping-templates/Mutation.createOrganization.request.vtl [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (brettstack) [inline mapping-templates/Mutation.createOrganization.request.vtl:1]: Are we using VTL instead of, say, a Lambda Function? Will this be too limiting?

Comment (bboure) [reply mapping-templates/Mutation.createOrganization.request.vtl:1]: With DynamoDb dataSources, vtl is the right choice for cost and performance (no cold start, no execution cost).

**Comment by sanathkr** [comment]
@

> @pipesorter I am not using Amplify, so I can't really tell. (I personally use the [Apollo Client](https://www.apollographql.com/docs/react/))
> 
> The good thing about Amplify, however, is that it handles offline support, Cognito, etc., out-of-the-box.
> I would definitely consider it for this project.

Do you think we can setup Amplify Lib without using their CLI? My understanding is the Amplify CLI will not be compatible with Serverless framework right?

**Comment by bboure** [comment]
> Do you think we can setup Amplify Lib without using their CLI? My understanding is the Amplify CLI will not be compatible with Serverless framework right?

AFAIK, Amplify-cli and Amplify lib are 2 standalone libraries. One is for deploying the infrastructure, the other one uses it.
There is no issue in using the library without the cli.

There might be a few gotchas like, I think the cli auto-generates code (models) for the library.
We might have to write those ourselves. 
Apart from that, no problem at all.

**Comment by bboure** [comment]
I just added more examples (lambda).

I noticed this repo is a fork of a [template project](https://github.com/wizeline/serverless-fullstack)

I haven't touched anything but I would suggest doing some cleanup and remove what we don't need.
For example I don't think you need express?

**Comment by sanathkr** [comment]
> I just added more examples (lambda).
> 
> I noticed this repo is a fork of a [template project](https://github.com/wizeline/serverless-fullstack)
> 
> I haven't touched anything but I would suggest doing some cleanup and remove what we don't need.
> For example I don't think you need express?

Good point. I don't think we need express here, but @brettstack  correct me if I am wrong.

**Comment by brettstack** [comment]
Right, we don't need Express. What's left before we can merge this?

**Review by brettstack — APPROVED** [review]

**Pull request merged**